Домашнее задание №2

geracimov opened the pull request.

geracimov committed:
1-4 tasks

geracimov committed:
5-6 tasks

geracimov committed:
CountMap implementation

geracimov committed:
CountMap implementation

Valeran86 commented on src/main/java/ru/sbt/collections/Counter1DifferentWords.java:
```diff
@@ -5,17 +5,148 @@
+    }
+
+    private static void printHeaderString() {
+        System.out.println("--- " + Thread.currentThread().getStackTrace()[2].getMethodName()
```
оригинально :)

geracimov replied:
Можно как то проще получить имя текущего выполняемого метода ?

Valeran86 commented on src/main/java/ru/sbt/collections/Counter1DifferentWords.java:
```diff
@@ -5,17 +5,148 @@
+        String[] arr = new String[list.size()];
+        arr = list.toArray(arr);
+        Collections.addAll(resultSet, arr);
+        resultSet.remove("");
```
можно при задать -1 `split( REGEXP_SPLIT_STRING, -1 )` и будет тот же эффект

Valeran86 commented on src/main/java/ru/sbt/collections/Counter1DifferentWords.java:
```diff
@@ -5,17 +5,148 @@
+            // если включить проверку на равенство ниже, то получим список слов с уникальной длинной,
+            // в противном случае получаем список всех строк из вх множества words, отсортированных по длинне строки
+            //if (o1.length() != o2.length())
+            ret = o1.length() > o2.length() ? 1 : -1;
```
список уникальных слов или вообще всех слов с повторами?

geracimov replied:
Список уникальных, на вход в метод же передается set

Valeran86 commented on src/main/java/ru/sbt/collections/Counter1DifferentWords.java:
```diff
@@ -5,17 +5,148 @@
+    public static void printWordsFrequency(List<String> words) {
+        HashMap<String, Integer> wc = new HashMap<>();
+        for (String word : words) {
+            wc.put(word, wc.containsKey(word) ? wc.get(word) + 1 : 1);
```
стоит использовать `merge` вместо `put`

geracimov replied:
изменил

Valeran86 commented on src/main/java/ru/sbt/collections/Counter1DifferentWords.java:
```diff
@@ -5,17 +5,148 @@
+     * Задание 4: Выведите на экран все строки файла в обратном порядке.
+     */
+    public static void printLinesReverse(List<String> strings) {
+        Collections.reverse(strings);
```
хороший однострочный вариант!

geracimov committed:
- проведен рефактринг: все public методы унифицированы принимают на вход String
- изменен printWordsFrequency, put заменен на merge
- упрощен метов getUniqWords
- добавлена проверка на null для вх параметров
- изменен стиль кода (пробелы)

CountMap implementation
- Изнемены несколько методов изходя из рекомендаций прерыдущего задания

CountMapUtils
- Реализованы методы с СComparable и Comparator

Valeran86 commented on src/main/java/ru/sbt/collections/CollectionUtils.java:
```diff
@@ -3,159 +3,155 @@
+    public static <T extends Comparable<T>> List<? super T> range( List<? extends T> list, T min, T max ) {
+        if ( list == null || min == null || max == null ) return null;
+
+        Collections.sort( list );
```
как сортировка влияет на отбор элементов? разве в цикле ниже учитывается порядок элементов?
